### ISSUE TITLE
Feature request: Support for local files instead of commits

### ISSUE THREAD

**Issue by maxless** [gollum/gollum#1031]
If you're working in an external text editor, you have to commit to see any change even if you're trying to do something small. Not to mention you have to constantly squash commits or have a lot of them around. It would be great if you could use Gollum to preview all changes before committing.

**Comment by SkyCrawl** [comment]
Great idea! I had this issue with external editors myself...

@dometto @bartkamphorst 
How about a couple of URL parameters? E.g.:
*  `/wiki/my-page?cwd` (display page from the working directory)
*  `/wiki/my-page?index`(display page from the git index)
*  `/wiki/my-page` (display page from the latest commit)
*  `/wiki/my-page?commit=...` (display page from the given commit)

**Comment by dometto** [comment]
IMHO this would be too much trouble to implement, and goes a bit against gollum's philosophy. Gollum offers the livepreview option to edit files and preview the changes. I think our efforts should go towards making the livepreview and editor better. See #826.

Of course I see how a markdown preview app would come in handy when working with external editors, but that's not limited to gollum, and I don't think gollum should solve it. Aren't there any other apps (browser-based or otherwise) that take care of this?

@sunny, what's your take on this?

**Comment by SkyCrawl** [comment]
> IMHO this would be too much trouble to implement

I think the only problem is git index :). All the other options are more or less ready/implemented. Integration itself is pretty easy. All of the above could even be squashed into a single parameter, despite personally, I don't think it's a good idea.

> Gollum offers the livepreview option to edit files and preview the changes. I think our efforts should go towards making the livepreview and editor better.

Same here. However, I don't think we should rule out convenient and trivial support for external editors so easily. I had the impression that Gollum didn't impose its live-preview editor on people? Actually, external editors are often much better, honed and more actively developed :). Also, users may be using Gollum only for a fraction of their (e.g.) markdown documents. What's wrong about using one's favourite editor for all (e.g.) markdown documents, if the user desires so? **On top of that, implementing `/wiki/my-page?cwd` (which is what @maxless requests) is, to my understanding, fairly easy :).**

> Of course I see how a markdown preview app would come in handy when working with external editors, but that's not limited to gollum, and I don't think gollum should solve it. Aren't there any other apps (browser-based or otherwise) that take care of this?

I think most editors have their own preview features, whatever the quality. Some apps seem more or less dedicated to previewing such things, for example Marked 2. However, what good are those preview features/apps for Gollum-specific pages? The best is, of course, Gollum itself :). And it's a single browser refresh away (with no need to repeatedly commit insignificant changes). Some people may be using git/Gollum for work and need to keep the history clean.

Although he doesn't use Gollum, my dad is a wonderful example of that. And even though he doesn't use Gollum, he's forced to use several external editors because none of them offers everything he needs :). In the end, I only have one voice but still - personally, I wouldn't wish for Gollum to "make more mess in this".

**Comment by sunny** [comment]
I think this is a good idea. I always thought that git for history was awesome but not being able to see what the file looks like without committing was annoying. I usually commit, amend, and end up messing up history.

Perhaps this argument should simply be called `?preview` and reflect what is on disk and not in the index.

**Comment by joshua-paddle** [comment]
Agreed. This is particularly useful to allow people to view how a page will look or interact with other pages, without having to commit. If I'm writing a reasonably large document, I really prefer to be able to use a text editor that provides all the advanced functionality (hotkeys, frequent saving, etc) that I'm used to, while still being able to easily view the generated page exactly as it will ultimately appear.

**Comment by dometto** [comment]
Ok, I yield. :) Dont know hoe feasible it is to get files from the git index (not sure if e.g. rugged supports that), and anyway that would require first git-adding new files before you could preview them. Wouldnt it be simpler to just get the file content directly from the filesystem. For safety reasons we should only display files under the git directory, and there should be a switch to enable this feature.

PRs very welcome, if anyone wants to try!

**Comment by SkyCrawl** [comment]
> Wouldnt it be simpler to just get the file content directly from the filesystem.

Yes, definitely :). I just mentioned the `index` option for the sake of completeness, not really thinking about it too much... I see now that it may be rather pointless.

**Comment by adius** [comment]
I would love to see that implemented!

**Comment by dirtyhenry** [comment]
👍 for this feature :)

**Comment by steve-bate** [comment]
+1

**Comment by markmoser** [comment]
This guard workflow works great for me.
https://gist.github.com/markmoser/f668ed9fb7fd63c5c0453cd39fcb9530

**Comment by vitalije** [comment]
I have added a new URL /tip/* which will serve pages and files directly from a working folder. That enables preview of not yet committed changes as requested in issue #1031.

The code is in my fork of this project https://github.com/vitalije/gollum
and I have posted a pull request.

Vitalije

PS: ruby is not my language of choice, therefore I didn't install any of dependencies and didn't run tests but the code is working for me and it solves #1031.

**Comment by dometto** [comment]
It just occurred to me that if we implement this feature, we should make it disabled by default and provide a `--local-preview` switch to enable it. Public gollum instances shouldn't by default give access to the server's filesystem..

**Comment by bartkamphorst** [comment]
> It just occurred to me that if we implement this feature, we should make it disabled by default and provide a --local-preview switch to enable it.

Agreed! I'm still on the fence about this feature in general, but apparently many people like the idea. 

Instead of adding this to the 5.0 milestone, perhaps we postpone implementing this until 5.1?

**Comment by dometto** [comment]
> Agreed! I'm still on the fence about this feature in general, but apparently many people like the idea.

Same general feeling. It appears to be one of the most requested features though, and we could keep it pretty barebones:

* add a '/preview/<caputure>' route
* see if <caputure> is a valid file under the repo root
  * but only if: `--local-preview` is enabled and the repository is not bare
* read the file and set a preview page, just like for the normal `preview` route
* clearly mark that the file is the version in the working directory, not the committed version
* no editing, uploading, etc. possible
* no way to browse the filesystem from the app, so it's only for people who use a different editor and want to have a way to preview

Ok with postponing until 5.1. If anyone would like to see this feature sooner, feel free to work on a PR against the `5.x` branch.

**Comment by blakewatson** [comment]
Just wanted to add my 👍 for this feature.

**Comment by thorlucas** [comment]
+1

I’d use Gollum to view my personal wiki as I edit it in nvim / vimwiki. Since I take notes directly during lecture, having to push commits etc is entirely ruled out as it interrupts workflow.

**Comment by thorlucas** [comment]
Whats wrong with this? https://github.com/vitalije/gollum/pull/1

**Comment by dometto** [comment]
@thorlucas just some quick notes on that solution:

* it's from 2016 so probably some things have changed in the gollum API since then (e.g. how to instantiate a preview page).
* It just assumes markdown is the default language (gollum allows users to customize this)
* there's no protection against directory traversal (i.e. someone could use `/tip` to read out `/etc/passwd`)
* there's no switch to turn this feature on or off, as discussed earlier in this thread

We'd consider a PR that addresses those issues, but safety first -- gollum is not used only locally, but also a server/web app.

**Comment by rdxmb** [comment]
we use mkdocs for things like that ...

**Comment by sarnobat** [comment]
> we use mkdocs for things like that ...

I love mkdocs apart from the fact that it doesn't support mediawiki. I don't want to migrate 17 years of exobrain notes to it :(

+1 begging for this feature.
